검색 기능 추가

pear96 opened the pull request:
- 키워드와 일기의 제목, 내용이 일치하는지 검색
- 일기의 작성 날짜가 특정 범위 안에 존재하는지 검색
- 일기의 감정에 따라 검색
-> 전부 페이지네이션 적용
-> 사용자 id 기반으로 검색하는 것이라 user Id에 index 적용

close #12 #25 #26 #27

pear96 committed:
refactor: DiaryRepository 분리

- DiaryRepository : 기본 CRUD 자동 생성하는 Spring Data 인터페이스
- DiaryCustomRepository : Mongo Template으로만 처리 가능한 커스텀 쿼리를 처리하는 메서드들의 목록이 담긴 인터페이스
- DiaryCustomRepositoryImpl : 커스텀 인터페이스를 직접 구현한 클래스
- DiaryRepository가 DiaryCustomRepository를 extends해서 DiaryCustomRepositoryImpl이 자동으로 빈으로 등록된다.

pear96 committed:
feat: 일기 제목, 내용으로 검색 #25

- Full Text 검색을 Mongo Text Index로 일단 구현하고, 추후 elastic search로 변경할 것을 고려해 `DiarySearchService`를 인터페이스로 만들어둠.
- `MongoTextSearchService`는 Mongo Text Index로 검색한다.
- diary.search.engine의 값에 따라 mongo로 할지 elastic으로 할지 설정해 두었다.
- MySQL 연결 제거

pear96 committed:
feat: 일기 제목, 내용으로 검색 #25

- 사용자 id 검색 조건 추가

pear96 committed:
feat: 날짜 검색, 감정 검색 추가 및 pagination 적용

- `/diary` -> `/diaries` 로 변경
  -  RESTful URL 설계에서 자원의 루트는 복수형 (“diaries”)으로 표기하는 것이 일반적이야.
- /keyword : 키워드 기반 일기 title, content 검색 #25
- /date : start, end로 날짜 범위 검색 (end 없으면 start로 설정) #26
- /emotion : 감정 필터링 #27
- 전체 조회 및 검색 관련 api 전부 페이지네이션 적용
- 전역 페이지네이션 설정을 통해 page size가 제한없이 커지는 것을 막음

pear96 committed:
refactor: exception

- BAD REQUEST로 응답해야하는 예외들 묶어서 처리

pear96 committed:
feat: emotion request param 처리

- request param에 emotion 들어올 때 역직렬화 추가

pear96 committed:
feat: Diary에 userId 인덱스 추가

- 사용자가 작성한 일기 찾을 때 전부 사용자 id 기반으로 조회하기 때문에 인덱스 적용

pear96 committed:
chore: 인덱스 스크립트 추가

pear96 committed:
chore: 인덱스 스크립트 추가

kochungcheon commented on diary-api/src/main/java/chungha/diary/repository/DiaryCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,79 @@
+@Repository
+public class DiaryCustomRepositoryImpl implements DiaryCustomRepository {
+	private final MongoTemplate mongoTemplate;
+	private static final String USER_ID = "userId";
```
스네이크케이스 고려해보는 건 어떨까?
mongoDB 에 다른 필드들은 다 스네이크 케이스로 필드명이 저장되었더라구

pear96 replied:
허거걱 이거 지금 바꿔도 괜찮을랑가 모르겠다...

kochungcheon commented on diary-api/src/main/java/chungha/diary/repository/DiaryRepository.java:
```diff
@@ -1,67 +1,8 @@
-		mongoTemplate.remove(new Query(Criteria.where("_id").is(diaryId)), Diary.class);
-	}
 
+public interface DiaryRepository extends MongoRepository<Diary, String>, DiaryCustomRepository {
```
DiaryCustomRepository에 대응되는 imple을 만든 이유가 궁금합니다.

pear96 replied:
질문을 이해를 못 하겠어요ㅠ DiaryCustomRepositoryImpl 이 아니라 DiaryRepositoryImpl이요,,?

kochungcheon replied:
수정했습니다

pear96 replied:
DiaryCustomRepositoryImpl에서 실제로 제가 필요한 쿼리들을 mongoTemplate으로 수행하는 메서드들이 구현되어있는데
DiaryRepository가 DiaryCustomRepository를 extends 하면

Spring Data MongoDB는 내부적으로 DiaryCustomRepositoryImpl을 자동으로 감지하고
이를 DiaryRepository의 구현체에 결합하여 하나의 Bean으로 등록합니다.

DiaryService에서 DiaryRepository만 의존해도 DiaryCustomRepositoryImpl의 메서드들을 쓸 수 있어서 그렇습니다.

이거를 설명했다고 생각해서 사실 질문이 이해가 안됐는데
DiaryCustomRepository 인터페이스만 쓰면 되는거 아니냐 이런 말은 아닌거쥬??

kochungcheon reviewed (commented):
이번 PR 도 고생많으셨습니다. 디코를 통해 충분히 설명된 코드들이 많았지만, 리뷰 몇 개 남기고 갑니다.

kochungcheon commented:
> 이번 PR 도 고생많으셨습니다. 디코를 통해 충분히 설명된 코드들이 많았지만, 리뷰 몇 개 남기고 갑니다.

날짜 다루는 것 보며 든 생각인 데, 2038 문제에 저희 시스템은 안전한지 한 번 확인해보고 싶네요
https://ko.wikipedia.org/wiki/2038%EB%85%84_%EB%AC%B8%EC%A0%9C

pear96 committed:
fix: Diary의 userId 스네이크 케이스 적용

- `Diary` entity의 `userId`에 `@Field("user_id")` 로 MongoDB에 `user_id` 로 저장되도록 변경
- `DiaryCustomRepositoryImpl`에서 사용자 아이디로 우선 일기를 검색할 때 `userId` 값에서 `user_id` 로 변경

pear96 commented:
comment에 어떻게 답댓 다는지 모르겠네요ㅠㅠ 일단 태그,,
@kochungcheon 

# Java time 패키지의 [LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html#MAX)

> The maximum supported LocalDateTime, '+999999999-12-31T23:59:59.999999999'. This is the local date-time just before midnight at the end of the maximum date. This combines [LocalDate.MAX](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html#MAX) and [LocalTime.MAX](https://docs.oracle.com/javase/8/docs/api/java/time/LocalTime.html#MAX). This could be used by an application as a "far future" date-time.

연/월/일 을 각각 int/short/short 로 숫자로 따로 저장하고 있기 때문에, 1970년부터 초 단위로 센 값 기준으로 저장하지 않음. 저장 방식이 다름. 따라서 2038과 상관없음


# MongoDB의 [Date](https://www.mongodb.com/ko-kr/docs/manual/reference/bson-types/#date)

> BSON Date는 Unix epoch(1970년 1월 1일) 이후의 밀리초 수를 나타내는 64비트 정수입니다. 그 결과 과거와 미래까지 약 2억 9천만년의 표현 가능한 날짜 범위가 생성됩니다.

<img width="661" alt="image" src="https://github.com/user-attachments/assets/92a06002-2441-4381-a32d-6f60cb738b58" />
